packages/asset-swapper/src/network/source_sampler.ts

```diff
@@ -0,0 +1,214 @@
+    TSellSamplerFunctionName = keyof TSellSamplerContract,
+    TBuySamplerFunctionName = keyof TBuySamplerContract
+> {
+    chain: Chain,
+    sellSamplerContractArtifactName: string;
+    buySamplerContractArtifactName: string;
+    sellSamplerContractType: ContractWrapperType<TSellSamplerContract>,
+    buySamplerContractType: ContractWrapperType<TBuySamplerContract>,
+    sellContractSellFunctionName: TSellSamplerFunctionName,
+    buyContractBuyFunctionName: TBuySamplerFunctionName,
+}
+
+export interface SamplerEthCall<TFillData, TSamplerFunctionArgs extends any[], TSamplerFunctionReturn> {
+    args: TSamplerFunctionArgs;
+    getDexSamplesFromResult(result: TSamplerFunctionReturn): DexSample<TFillData>[];
+}
+
+export type UnwrapContractFunctionReturnType<T> = T extends ContractFunctionObj<infer U> ? U : T;
+
+// Base class for a standard sampler with on-chain quote functions.
+export abstract class OnChainSourceSampler<
+    TSellSamplerContract extends GeneratedContract,
+    TBuySamplerContract extends GeneratedContract,
+    TSellSamplerFunction extends ContractFunction<TSellSamplerFunctionArgs, TSellSamplerFunctionReturn>,
+    TBuySamplerFunction extends ContractFunction<TBuySamplerFunctionArgs, TBuySamplerFunctionReturn>,
+    TFillData extends FillData,
+    TSellSamplerFunctionName extends keyof TSellSamplerContract = keyof TSellSamplerContract,
+    TBuySamplerFunctionName extends keyof TBuySamplerContract = keyof TBuySamplerContract,
+    TSellSamplerFunctionArgs extends any[] = Parameters<TSellSamplerFunction>,
+    TBuySamplerFunctionArgs extends any[] = Parameters<TBuySamplerFunction>,
+    TSellSamplerFunctionReturn = UnwrapContractFunctionReturnType<ReturnType<TSellSamplerFunction>>,
+    TBuySamplerFunctionReturn = UnwrapContractFunctionReturnType<ReturnType<TBuySamplerFunction>>,
+    TOpts extends OnChainSourceSamplerOptions<
+        TSellSamplerContract,
+        TBuySamplerContract,
+        TSellSamplerFunctionName,
+        TBuySamplerFunctionName
+    > = OnChainSourceSamplerOptions<
+        TSellSamplerContract,
+        TBuySamplerContract,
+        TSellSamplerFunctionName,
+        TBuySamplerFunctionName
+    >
+> extends SourceSamplerBase {
+    protected readonly _chain: Chain;
+    protected readonly _sellContract: TSellSamplerContract;
+    protected readonly _buyContract: TBuySamplerContract;
+    protected readonly _sellContractHelper: ContractHelper<TSellSamplerContract>;
+    protected readonly _buyContractHelper: ContractHelper<TBuySamplerContract>;
+    protected readonly _sellContractFunction: TSellSamplerFunction;
+    protected readonly _buyContractFunction: TBuySamplerFunction;
+
+    protected constructor(opts: TOpts) {
+        super();
+        this._chain = opts.chain;
+        [this._sellContract, this._sellContractHelper] = createContractWrapperAndHelper(
+            opts.chain,
+            opts.sellSamplerContractType,
+            opts.sellSamplerContractArtifactName,
```

Comment:
It doesn't exist yet at this point, but maybe something like `` `/^(.+)Wrapper$/\`.exec(opts.sellSamplerContractType.name)[1]``?

Reply:
I mean like add `contractName: string` to the `ContractWrapperType<T>` interface, since abi-gen gives each contract wrapper class a `public static contractName`

Reply:
ahh yeah nice